### PR TITLE
Scrubbing online learners

### LINKED ISSUE #2614: Scrubbing OnlineLinear learners.
Sub task of https://github.com/dotnet/machinelearning/issues/2613
`AveragePerceptron`
`LinearSVM`
`OnlineGradientDescentTrainer`
and their base classes

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
fixes #2614

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/AveragedPerceptronWithOptions.cs [hunk 1]
```diff
@@ -25,7 +25,7 @@ public static void Example()
             {
                 LossFunction = new SmoothedHingeLoss.Options(),
                 LearningRate = 0.1f,
-                DoLazyUpdates = false,
+                LazyUpdates = false,
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/AveragedPerceptronWithOptions.cs]:
```suggestion
                LazyUpdate = false,
```

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/AveragedPerceptronWithOptions.cs [hunk 2]
```diff
@@ -25,7 +25,7 @@ public static void Example()
...
                 RecencyGain = 0.1f,
                 NumberOfIterations = 10
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/AveragedPerceptronWithOptions.cs]:
```suggestion
                MaximumNumberOfIterations = 10
```

File: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedLinear.cs [hunk 3]
```diff
@@ -58,15 +58,15 @@ public abstract class AveragedLinearOptions : OnlineLinearOptions
         /// Default is <see langword="true" />.
         /// </value>
         [Argument(ArgumentType.AtMostOnce, HelpText = "Instead of updating averaged weights on every example, only update when loss is nonzero", ShortName = "lazy")]
-        public bool DoLazyUpdates = true;
+        public bool LazyUpdates = true;
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/Online/AveragedLinear.cs]:
```suggestion
        public bool LazyUpdate = true;
```

File: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs [hunk 4]
```diff
@@ -126,23 +126,23 @@ internal AveragedPerceptronTrainer(IHostEnvironment env, Options options)
         /// <param name="featureColumnName">The name of the feature column.</param>
         /// <param name="learningRate">The learning rate. </param>
         /// <param name="decreaseLearningRate">Whether to decrease learning rate as iterations progress.</param>
-        /// <param name="l2RegularizerWeight">L2 Regularization Weight.</param>
+        /// <param name="l2Regularization">Weight of L2 regularization term.</param>
         /// <param name="numIterations">The number of training iterations.</param>
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs]:
```suggestion
        /// <param name="maximumNumberOfIterations">The number of training iterations.</param>
```

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs [hunk 5]
```diff
@@ -95,22 +95,22 @@ public override LinearRegressionModelParameters CreatePredictor()
         /// <param name="featureColumn">Name of the feature column.</param>
         /// <param name="learningRate">The learning Rate.</param>
         /// <param name="decreaseLearningRate">Decrease learning rate as iterations progress.</param>
-        /// <param name="l2RegularizerWeight">L2 Regularization Weight.</param>
+        /// <param name="l2Regularization">Weight of L2 regularization term.</param>
         /// <param name="numIterations">Number of training iterations through the data.</param>
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]:
```suggestion
        /// <param name="maximumNumberOfIterations">Number of training iterations through the data.</param>
```
Please also check other online learners' APIs.

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs [hunk 6]
```diff
@@ -95,22 +95,22 @@ public override LinearRegressionModelParameters CreatePredictor()
...
         /// <param name="lossFunction">The custom loss functions. Defaults to <see cref="SquaredLoss"/> if not provided.</param>
         internal OnlineGradientDescentTrainer(IHostEnvironment env,
```
Comment (wschin) [inline src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]:
```suggestion
        internal OnlineGradientTrainer(IHostEnvironment env,
```
A descent algorithm ensures the decrease of function value per iteration. However, this is not true for most stochastic gradient learners.

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs (hunk 5 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: I don't see any stop rule in `OnlineLinearTrainer`.`TrainCore` but maybe it just me.

---
In reply to: [263972702](https://github.com/dotnet/machinelearning/pull/2892#discussion_r263972702) [](ancestors = 263972702)

Comment (wschin) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: I am too lazy to dig into the code, but my suggestion just reflects this parameter's description.

---
In reply to: [263975389](https://github.com/dotnet/machinelearning/pull/2892#discussion_r263975389) [](ancestors = 263975389,263972702)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs (hunk 6 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: I don't get your comment at all.

---
In reply to: [263972819](https://github.com/dotnet/machinelearning/pull/2892#discussion_r263972819) [](ancestors = 263972819)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/AveragedPerceptronWithOptions.cs (hunk 2 above)
Comment (Ivanidzo4ka) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/AveragedPerceptronWithOptions.cs]: No, It's number of iterations.

---
In reply to: [263972500](https://github.com/dotnet/machinelearning/pull/2892#discussion_r263972500) [](ancestors = 263972500)

File: src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs (hunk 5 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.StandardLearners/Standard/Online/OnlineGradientDescent.cs]: But you suggest same description as I have right now....

---
In reply to: [264346723](https://github.com/dotnet/machinelearning/pull/2892#discussion_r264346723) [](ancestors = 264346723,263975389,263972702)

File: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs (hunk 4 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.StandardLearners/Standard/Online/AveragedPerceptron.cs]: I don't get your suggestion.

---
In reply to: [263972642](https://github.com/dotnet/machinelearning/pull/2892#discussion_r263972642) [](ancestors = 263972642)

**Comment by abgoswam** [comment]
>         /// The inexactness tolerance for averaging.

not sure what this means .. do we just mean tolerance ?

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/Online/AveragedLinear.cs:102 in 55c4eda. [](commit_id = 55c4eda558e5e1633fe5df238522f1c6718ce586, deletion_comment = False)
